### ISSUE TITLE
implement curve25519-sha256@libssh.org

### ISSUE THREAD

**Issue by mfazekas** [net-ssh/net-ssh#214]
see #211

**Comment by mfazekas** [comment]
http://git.libssh.org/users/aris/libssh.git/plain/doc/curve25519-sha256@libssh.org.txt?h=curve25519
https://stribika.github.io/2015/01/04/secure-secure-shell.html

@chr4 is this something you'd interested and have time to work on?

**Comment by Seegras** [comment]
That is not the only algorithm missing (but the only relevant kex that is missing. The others use NIST curves, which you should avoid anyway). 

MACs: 
umac-64@openssh.com
umac-128@openssh.com
hmac-sha1-etm@openssh.com
hmac-sha1-96-etm@openssh.com
hmac-sha2-256-etm@openssh.com
hmac-sha2-512-etm@openssh.com
hmac-md5-etm@openssh.com
hmac-md5-96-etm@openssh.com
hmac-ripemd160-etm@openssh.com
umac-64-etm@openssh.com
umac-128-etm@openssh.com
Basically, you can ignore every MAC with md5 and sha1

Ciphers:
chacha20-poly1305@openssh.com

**Comment by odigity** [comment]
+1

Just ran into this today, and took me a long time to figure out what the problem was, because I could ssh to user@host, but not cap to it...

**Comment by kdfreeheart** [comment]
+1

**Comment by mfazekas** [comment]
@odigity ed25519 for private key, right? what was the error? 2.9.2 should give meaningfull errors about this. And yes there are many stuff to implement, PR-s welcome.

**Comment by odigity** [comment]
Contents of `config/deploy/development.rb`:

    server 'web1.porcclub.dev', roles: ['web'], ssh_options: {
      auth_methods:  ['publickey'],
      forward_agent: true,
      keys:          [ENV['HOME'] + '/.ssh/porcclub_deploy_rsa'],
    #  keys:          [ENV['HOME'] + '/.ssh/porcclub_deploy_ed25519'],
      user:          'portal'
    }

With RSA key:

    vagrant@web1:~/portal$ cap development check_ssh_forwarding
    DEBUG [3f665050] Running /usr/bin/env whoami on web1.porcclub.dev
    DEBUG [3f665050] Command: /usr/bin/env whoami
    DEBUG [3f665050] Finished in 0.087 seconds with exit status 0 (successful).
    DEBUG [3f665050] 	portal
    DEBUG [3f665050] Finished in 0.087 seconds with exit status 0 (successful).

Comment our RSA key, uncomment ed25519 key:

    vagrant@web1:~/portal$ cap development check_ssh_forwarding
    DEBUG [834e3868] Running /usr/bin/env whoami on web1.porcclub.dev
    DEBUG [834e3868] Command: /usr/bin/env whoami
    (Backtrace restricted to imported tasks)
    cap aborted!
    SSHKit::Runner::ExecuteError: Exception while executing on host web1.porcclub.dev: Authentication failed for user portal@web1.porcclub.dev
    /home/vagrant/portal/lib/capistrano/tasks/misc.rake:6:in `block (2 levels) in <top (required)>'
    Net::SSH::AuthenticationFailed: Authentication failed for user portal@web1.porcclub.dev
    /home/vagrant/portal/lib/capistrano/tasks/misc.rake:6:in `block (2 levels) in <top (required)>'
    Tasks: TOP => check_ssh_forwarding
    (See full trace by running task with --trace)

Confirmed that both public keys are in portal user's `authorized_keys` file, and am able to SSH on the command line using either key.

**Comment by odigity** [comment]
The good news is I managed to find a way around this problem.

Turns out if you're using SSH agent forwarding, rather than having net-ssh source the key directly, then ed25519 keys work fine under Capistrano.

I did the following, starting on my Ubuntu 14.04.1 desktop:

1. I killed gnome-keyring, which comes on by default in Ubuntu and ALSO doesn't support ed25519.
2. Started up ssh-agent instead.
3. Added my ed25519 key to add via `ssh-add`.
4. Enabling forwarding in my Vagrantfile via `config.ssh_forward_agent = true` and reloaded the VM.
5. vagrant ssh
6. cap development check_ssh_forwarding

And it worked.  The auth chain extended into the next hop (portal@web.porcclub.dev), and from there to my GitLab server (both places had the ed25519 pub key already configured, obviously).

**Comment by odigity** [comment]
Update:  All of a sudden it's no longer working, but RSA still is.  Not sure if I made a mistake earlier, or if I screwed something up after.

Guess I have to fallback to RSA keys for now after all...

**Comment by mfazekas** [comment]
ed25519 is coming thanks to https://github.com/cryptosphere/rbnacl see #228
no password encryption/agent support yet

**Comment by odigity** [comment]
Woo hoo!

**Comment by centromere** [comment]
+1

**Comment by yolk** [comment]
Any news on this?

**Comment by doits** [comment]
+1, stumbled at this right now, capistrano (using this library) was always asking for ssh password and I had to debug this for > 20 min to find out it was because of me using an ed25519 key ...

**Comment by indiv0** [comment]
+1, the vagrant-libvirt plugin uses this library and because of this issue it ends up asking for an SSH password if an ed25519 key is used.

**Comment by JamesHarrison** [comment]
+1 here - if it's not going to be fixed, can I suggest that the unimplemented key error be surfaced so that this isn't utterly opaque without the ssh_options verbose flag set to debug?

**Comment by mfazekas** [comment]
@JamesHarrison can you please clarify the usecase?
You have ed25519 key set up and net-ssh fails back to plain password auth, and you want some warning explaining that ed25519 key was ignored as it's non supported?

**Comment by JamesHarrison** [comment]
@mfazekas Apologies, failed to spot this wasn't the Capistrano git, my mistake! In Capistrano, if a server is configured not to permit password authentication then you simply get an authentication failed message with no explanation as to what happened or why your authentication failed. Capistrano doesn't display net-ssh messages by default, even if they're errors.

**Comment by qzio** [comment]
:+1:
I was struck by this using chef's `knife ssh`.

Would really nice to at least get an error.

**Comment by ernetas** [comment]
:+1: Really missing this.

P. S. @qzio not sure about Chef, but e.g., in Capistrano you can increase verbosity of SSH and that shows a normal error.

**Comment by mfazekas** [comment]
FYI: just released net-ssh 4.0.0.alpha1 with support for ed25519

**Comment by jkraemer** [comment]
this works great for me with alpha1 and alpha3 but appears to be broken again in alpha4 and beta1.

alpha4 yields `NameError: uninitialized constant Net::SSH::Buffer::ED25519`, beta1 yields `uninitialized constant Net::SSH::Authentication::ED25519Loader::ED25519_LOAD_ERROR` when trying to connect using an ed25519 key. Ruby 2.3.0.

**Comment by mfazekas** [comment]
Can you exectute to debug this problem?
```
bundle exec irb
require 'net/ssh'
require 'net/ssh/authentication/ed25519_loader'
puts Net::SSH::Authentication::ED25519Loader::ERROR
```

**Comment by jkraemer** [comment]
Sure. Here's the result with beta1:

```
irb(main):001:0> require 'net/ssh'
=> true
irb(main):002:0> require 'net/ssh/authentication/ed25519_loader'
=> false
irb(main):003:0> puts Net::SSH::Authentication::ED25519Loader::ERROR
rbnacl-libsodium is not part of the bundle. Add it to Gemfile.
=> nil
```


-- 
Jens Kr�mer
https://jkraemer.net/

**Comment by mfazekas** [comment]
@jkraemer i've just updata 4.0.0-beta2 which should fix raise the correct error. As the error stated you need to add rbnacl-libsodium and rbnacl to your gemfile in order for ed25519 support to work.

**Comment by jkraemer** [comment]
thanks, this works!

On 08/08, Mikl�s Fazekas wrote:
> @jkraemer i've just updata 4.0.0-beta2 which should fix raise the correct error. As the error stated you need to add rbnacl-libsodium and rbnacl to your gemfile in order for ed25519 support to work.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/net-ssh/net-ssh/issues/214#issuecomment-238197008

-- 
Jens Kr�mer
https://jkraemer.net/

**Comment by epinault** [comment]
Any idea when this will be released in non beta?

**Comment by mfazekas** [comment]
@epinault have you tried beta did it work for you? I try to do an rc release this week

**Comment by epinault** [comment]
I did install the beta and it worked. Just was having trouble to pick it up with Bundler.

**Comment by mfazekas** [comment]
@epinault what's the issue with bundler? You should be able to do 

```ruby
gem 'net-ssh', '= 4.0.0.beta2'
```

or even

```ruby
gem 'net-ssh', github: 'net-ssh/net-ssh'
```

**Comment by epinault** [comment]
I was trying >= 4.0 and would not work because it is in beta . But I found that you have to be specific with beta version.  Anyway, not a big issue. Can't wait for the final release of this gem :)

**Comment by mfazekas** [comment]
https://rubygems.org/gems/net-ssh/versions/4.0.0.rc1 is now released. For now rbnacl-libsodium, rbnacl, bcrypt_pbkdf have to be added to Gemfile for ed25519 to work.

**Comment by leehambley** [comment]
With net-ssh 4.0 out (congratulations on a solid release 🍾 !) Capistrano (and SSHKit) have been weighing what to do about the new important dependencies over at https://github.com/capistrano/capistrano/pull/1825

We think it'd be logical to have net-ssh require these new gems directly as ed25519 keys are rapidly becoming the default.

What do you think? (Sorry if I'm hijacking this issue), I'm picking up mostly on the line 

>  For now rbnacl-libsodium, rbnacl, bcrypt_pbkdf have to be added to Gemfile for ed25519 to work.

And wondering what the grounds for the "for now" part are!

**Comment by mfazekas** [comment]
@leehambley  see #476
